# Gradient Boosting - Hyperparameter Tuning

The previous notebook explored selected Gradient Boosting hyperparameters manually. Increasing tree depth produced the largest improvement in fraud detection performance, but deeper trees also showed diminishing returns and an increasing risk of overfitting.

This notebook extends the manual experiments by introducing systematic hyperparameter tuning.

The tuning process will:

- explore multiple Gradient Boosting hyperparameter configurations,
- use F1-score as the primary optimization metric,
- preserve the temporal structure of the transaction data during validation,
- keep the 2018 validation set separate from hyperparameter tuning,
- keep 2019 completely unseen for final model evaluation.

## Experimental Design

The available labeled transaction history is divided according to time:

- **2010–2017:** model development and hyperparameter tuning
- **2018:** validation and model comparison
- **2019:** final out-of-time test set

Hyperparameter tuning must use only information available within the 2010–2017 development period. The temporal order of transactions will be preserved during cross-validation to avoid training models on future observations and evaluating them on the past.

In [63]:
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.ensemble import GradientBoostingClassifier

from sklearn.metrics import (
    recall_score,
    precision_score,
    f1_score,
    confusion_matrix,
)

from finsight.model_evaluation import evaluate_model

from finsight.fraud_preprocessing import (
    encode_categorical_features,
    handle_missing_values,
)

from finsight.database import (
    connect_to_database,
    test_database_connection,
)

from finsight.fraud_data import (
    BASE_FEATURES,
    download_yearly_data,
)

In [2]:
engine = connect_to_database()
test_database_connection(engine)

Connected!


## Development Data

Hyperparameter tuning is performed exclusively on data from 2010–2017.

The development dataset is loaded separately for each year to preserve temporal boundaries required for time-aware cross-validation. Non-fraud transactions are sampled proportionally to the original yearly transaction volumes, with a total sampling budget of 300,000 observations.

The previously selected base feature set is retained so that the tuning experiments isolate the effect of model hyperparameters rather than changes in feature representation.

In [3]:
features = BASE_FEATURES

yearly_data = download_yearly_data(
    engine,
    start_year=2010,
    end_year=2017,
    total_non_fraud_limit=300000,
    features=features
)

In [4]:
for year, data in yearly_data.items():
    print(
        year,
        data["X"].shape,
        data["y"].value_counts().to_dict()
    )

2010 (37148, 13) {False: 34575, True: 2573}
2011 (36048, 13) {False: 36011, True: 37}
2012 (37815, 13) {False: 36892, True: 923}
2013 (39124, 13) {False: 37787, True: 1337}
2014 (38803, 13) {False: 38139, True: 664}
2015 (40896, 13) {False: 38707, True: 2189}
2016 (41251, 13) {False: 38803, True: 2448}
2017 (39258, 13) {False: 39086, True: 172}


## Temporal Development Set

The yearly datasets are concatenated in chronological order from 2010 to 2017.

The feature matrix and target vector are kept separate, while an additional year vector is created to identify the temporal origin of each observation. The year information is used only for constructing temporal validation folds and is not included as a model feature.

In [8]:
X_development = pd.concat(
    [yearly_data[year]["X"] for year in range(2010, 2018)],
    ignore_index=True
)
X_development.shape

(310343, 13)

In [9]:
y_development = pd.concat(
    [yearly_data[year]["y"] for year in range(2010, 2018)],
    ignore_index=True
)

y_development.shape

(310343,)

In [10]:
development_years = []

for year in yearly_data:
    development_years.extend(
        [year] * len(yearly_data[year]["X"])
    )

development_years = pd.Series(development_years)

development_years.shape

(310343,)

In [11]:
development_years.value_counts().sort_index()

2010    37148
2011    36048
2012    37815
2013    39124
2014    38803
2015    40896
2016    41251
2017    39258
Name: count, dtype: int64

## Temporal Cross-Validation

In [12]:
train_mask = development_years.isin([2010, 2011, 2012])
train_indices = development_years[train_mask].index
train_indices

RangeIndex(start=0, stop=111011, step=1)

In [16]:
development_years[train_mask].unique()

array([2010, 2011, 2012])

In [14]:
val_mask = development_years.isin([2013])
val_indices = development_years[val_mask].index
val_indices

RangeIndex(start=111011, stop=150135, step=1)

In [17]:
development_years[val_mask].unique()

array([2013])

In [19]:
train_indices.intersection(val_indices)

RangeIndex(start=0, stop=0, step=1)

### Automated Temporal Folds

In [20]:
def create_temporal_folds(development_years, min_years=3):
    folds = []
    years = development_years.unique()

    for val_year in years[min_years:]:
        val_mask = (development_years == val_year)
        val_indices = development_years[val_mask].index

        train_mask = development_years < val_year
        train_indices = development_years[train_mask].index

        folds.append((train_indices, val_indices))
        
    return folds

In [22]:
folds = create_temporal_folds(development_years)
len(folds)

5

In [28]:
for train_indices, val_indices in folds:
    print(development_years[train_indices].unique(), development_years[val_indices].unique())

[2010 2011 2012] [2013]
[2010 2011 2012 2013] [2014]
[2010 2011 2012 2013 2014] [2015]
[2010 2011 2012 2013 2014 2015] [2016]
[2010 2011 2012 2013 2014 2015 2016] [2017]


### Expanding Window Validation

To preserve the temporal structure of the transaction data, model configurations are evaluated using expanding-window cross-validation rather than random cross-validation.

Each fold trains the model using all available historical years and validates it on the immediately following year:

- Fold 1: 2010–2012 → 2013
- Fold 2: 2010–2013 → 2014
- Fold 3: 2010–2014 → 2015
- Fold 4: 2010–2015 → 2016
- Fold 5: 2010–2016 → 2017

This setup simulates repeated model development over time while preventing future observations from being used to predict the past.

The folds store only the training and validation indices. The corresponding feature and target data are selected from the development dataset during model evaluation.

## Hyperparameter Tuning

In [60]:
fold_f1_scores = []

for train_indices, val_indices in folds:
    X_train = X_development.iloc[train_indices].copy()
    y_train = y_development.iloc[train_indices]

    X_val = X_development.iloc[val_indices].copy()
    y_val = y_development.iloc[val_indices]

    X_train["mcc_key"] = X_train["mcc_key"].astype(str)
    X_val["mcc_key"] = X_val["mcc_key"].astype(str)

    X_train, X_val = encode_categorical_features(X_train, X_val)
    X_train, X_val = handle_missing_values(X_train, X_val)

    gb = GradientBoostingClassifier(max_depth=6, random_state=42)
    gb.fit(X_train, y_train)

    y_pred = gb.predict(X_val)
    f1 = f1_score(y_val, y_pred)

    fold_f1_scores.append(f1)

fold_f1_scores

[0.766107678729038,
 0.8132045088566827,
 0.5488721804511278,
 0.848318462594372,
 0.036231884057971016]

## Temporal Validation Diagnostics

The expanding-window validation revealed substantial variation in model performance across validation periods:

- **2013:** F1 = 0.766
- **2014:** F1 = 0.813
- **2015:** F1 = 0.549
- **2016:** F1 = 0.848
- **2017:** F1 = 0.036

The model generalizes well to several validation periods, but performance deteriorates noticeably in 2015 and collapses almost completely in 2017.

Before proceeding with hyperparameter tuning, these periods will be investigated to determine whether the performance degradation is associated with changes in:

1. fraud prevalence,
2. model error patterns,
3. the characteristics of fraudulent transactions,
4. the distribution of features relevant to fraud detection.

The objective is not to optimize the model specifically for individual validation years, but to understand whether its performance is sensitive to temporal changes in the data.

In [61]:
yearly_fraud = pd.DataFrame({
    "year": development_years,
    "is_fraud": y_development
})

fraud_by_year = yearly_fraud.groupby("year")["is_fraud"].agg(
    transactions="count",
    frauds="sum",
    fraud_rate="mean"
)

fraud_by_year["fraud_rate"] *= 100

fraud_by_year

,transactions,frauds,fraud_rate
year,,,
2010,37148,2573,6.926349
2011,36048,37,0.102641
2012,37815,923,2.440830
2013,39124,1337,3.417340
2014,38803,664,1.711208
2015,40896,2189,5.352602
2016,41251,2448,5.934402
2017,39258,172,0.438127


In [64]:
fold_results = []

validation_years = [2013, 2014, 2015, 2016, 2017]

for (train_indices, val_indices), year in zip(folds, validation_years):
    X_train = X_development.iloc[train_indices].copy()
    y_train = y_development.iloc[train_indices]

    X_val = X_development.iloc[val_indices].copy()
    y_val = y_development.iloc[val_indices]

    X_train["mcc_key"] = X_train["mcc_key"].astype(str)
    X_val["mcc_key"] = X_val["mcc_key"].astype(str)

    X_train, X_val = encode_categorical_features(X_train, X_val)
    X_train, X_val = handle_missing_values(X_train, X_val)

    gb = GradientBoostingClassifier(max_depth=6, random_state=42)
    gb.fit(X_train, y_train)

    y_pred = gb.predict(X_val)

    tn, fp, fn, tp = confusion_matrix(y_val, y_pred).ravel()

    fold_results.append({
        "year": year,
        "precision": precision_score(y_val, y_pred),
        "recall": recall_score(y_val, y_pred),
        "f1": f1_score(y_val, y_pred),
        "tn": tn,
        "fp": fp,
        "fn": fn,
        "tp": tp
    })

fold_results = pd.DataFrame(fold_results)

fold_results

,year,precision,recall,f1,tn,fp,fn,tp
0,2013,0.934338,0.649215,0.766108,37726,61,469,868
1,2014,0.873702,0.760542,0.813205,38066,73,159,505
2,2015,0.438438,0.733668,0.548872,36650,2057,583,1606
3,2016,0.964119,0.757353,0.848318,38734,69,594,1854
4,2017,0.048077,0.029070,0.036232,38987,99,167,5


In [65]:
fold_results["actual_fraud_rate"] = (
    (fold_results["tp"] + fold_results["fn"]) /
    (fold_results["tn"] + fold_results["fp"] + fold_results["fn"] + fold_results["tp"])
) * 100

fold_results["predicted_fraud_rate"] = (
    (fold_results["tp"] + fold_results["fp"]) /
    (fold_results["tn"] + fold_results["fp"] + fold_results["fn"] + fold_results["tp"])
) * 100

fold_results[
    ["year", "actual_fraud_rate", "predicted_fraud_rate"]
]

,year,actual_fraud_rate,predicted_fraud_rate
0,2013,3.417340,2.374502
1,2014,1.711208,1.489576
2,2015,5.352602,8.956866
3,2016,5.934402,4.661705
4,2017,0.438127,0.264914


In [66]:
fraud_features = X_development.copy()

fraud_features["year"] = development_years.values
fraud_features["is_fraud"] = y_development.values

fraud_features = fraud_features[fraud_features["is_fraud"] == True]

fraud_feature_medians = fraud_features.groupby("year")[
    [
        "amount",
        "user_previous_avg_amount",
        "amount_vs_user_avg",
        "card_previous_avg_amount",
        "amount_vs_card_avg"
    ]
].median()

fraud_feature_medians

,amount,user_previous_avg_amount,amount_vs_user_avg,card_previous_avg_amount,amount_vs_card_avg
year,,,,,
2010,74.290,43.540,28.830,44.160,29.180
2011,99.890,59.670,40.550,58.000,40.720
2012,47.710,41.000,4.750,38.990,7.010
2013,86.110,41.500,43.900,41.170,47.230
2014,87.495,41.930,40.560,38.795,44.515
2015,88.790,41.070,45.730,39.340,46.800
2016,84.665,40.695,43.540,39.700,44.095
2017,29.495,43.460,-10.925,37.610,-7.270


In [67]:
nonfraud_features = X_development.copy()

nonfraud_features["year"] = development_years.values
nonfraud_features["is_fraud"] = y_development.values

nonfraud_features = nonfraud_features[
    nonfraud_features["is_fraud"] == False
]

nonfraud_feature_medians = nonfraud_features.groupby("year")[
    [
        "amount",
        "user_previous_avg_amount",
        "amount_vs_user_avg",
        "card_previous_avg_amount",
        "amount_vs_card_avg"
    ]
].median()

nonfraud_feature_medians

,amount,user_previous_avg_amount,amount_vs_user_avg,card_previous_avg_amount,amount_vs_card_avg
year,,,,,
2010,29.72,40.300,-9.85,39.29,-8.190
2011,29.24,40.330,-10.67,39.44,-8.740
2012,28.97,40.245,-10.43,39.00,-8.520
2013,29.36,40.440,-10.53,39.19,-8.680
2014,28.98,40.150,-10.56,38.70,-8.450
2015,28.84,39.960,-10.73,38.64,-8.770
2016,28.93,40.000,-10.85,38.72,-9.025
2017,28.42,39.880,-10.87,38.43,-8.950


In [68]:
fraud_use_chip_by_year = pd.crosstab(
    fraud_features["year"],
    fraud_features["use_chip"],
    normalize="index"
) * 100

fraud_use_chip_by_year

use_chip,Chip Transaction,Online Transaction,Swipe Transaction
year,,,
2010,0.000000,90.050525,9.949475
2011,0.000000,94.594595,5.405405
2012,0.000000,75.081257,24.918743
2013,0.000000,84.293194,15.706806
2014,0.000000,84.638554,15.361446
2015,8.725445,85.153038,6.121517
2016,7.761438,85.620915,6.617647
2017,98.837209,0.000000,1.162791


In [69]:
nonfraud_use_chip_by_year = pd.crosstab(
    nonfraud_features["year"],
    nonfraud_features["use_chip"],
    normalize="index"
) * 100

nonfraud_use_chip_by_year

use_chip,Chip Transaction,Online Transaction,Swipe Transaction
year,,,
2010,0.000000,10.672451,89.327549
2011,0.000000,10.827247,89.172753
2012,0.000000,11.034913,88.965087
2013,0.000000,11.490724,88.509276
2014,0.000000,11.610163,88.389837
2015,70.651303,12.318185,17.030511
2016,71.115635,12.140814,16.743551
2017,70.861690,12.127104,17.011206


In [70]:
train_indices, val_indices = folds[2]

X_train_2015 = X_development.iloc[train_indices].copy()
y_train_2015 = y_development.iloc[train_indices]

X_val_2015 = X_development.iloc[val_indices].copy()
y_val_2015 = y_development.iloc[val_indices]

X_train_2015["mcc_key"] = X_train_2015["mcc_key"].astype(str)
X_val_2015["mcc_key"] = X_val_2015["mcc_key"].astype(str)

X_train_2015_encoded, X_val_2015_encoded = encode_categorical_features(
    X_train_2015,
    X_val_2015
)

X_train_2015_encoded, X_val_2015_encoded = handle_missing_values(
    X_train_2015_encoded,
    X_val_2015_encoded
)

gb_2015 = GradientBoostingClassifier(max_depth=6, random_state=42)
gb_2015.fit(X_train_2015_encoded, y_train_2015)

y_pred_2015 = gb_2015.predict(X_val_2015_encoded)

In [72]:
false_positive_mask_2015 = (
    (y_val_2015.to_numpy() == False) &
    (y_pred_2015 == True)
)

false_positives_2015 = X_val_2015.loc[false_positive_mask_2015]

false_positives_2015["use_chip"].value_counts(normalize=True) * 100

use_chip
Chip Transaction      97.860963
Online Transaction     1.993194
Swipe Transaction      0.145843
Name: proportion, dtype: float64

In [73]:
nonfraud_chip_mask_2015 = (
    (y_val_2015.to_numpy() == False) &
    (X_val_2015["use_chip"].to_numpy() == "Chip Transaction")
)

nonfraud_chip_predictions_2015 = y_pred_2015[nonfraud_chip_mask_2015]

nonfraud_chip_predictions_2015.mean() * 100

np.float64(7.360953669506711)

### 2015 Performance Degradation

The 2015 validation period shows a different failure mode from the severe performance collapse observed in 2017.

Although the model maintained relatively strong fraud detection ability in 2015 (**recall = 0.734**), precision decreased substantially to **0.438**, resulting in an F1-score of **0.549**. The primary source of this degradation was a large increase in false positives: **2,057 legitimate transactions were incorrectly classified as fraud**.

Further analysis revealed a major temporal change in the transaction channel distribution. Until 2014, legitimate transactions contained virtually no `Chip Transaction` observations. In 2015, however, **70.65% of legitimate transactions were chip-based**.

The model, trained only on earlier periods, therefore encountered a transaction channel that was largely absent from its historical training data.

This shift was strongly associated with the observed false-positive problem:

- **97.86% of all false positives in 2015 were Chip Transactions**
- **7.36% of legitimate Chip Transactions were incorrectly classified as fraud**
- fraud recall nevertheless remained relatively high at **0.734**

Interestingly, the distribution of legitimate transaction channels remained almost unchanged in 2016, while model performance recovered to **F1 = 0.848**. Under the expanding-window validation scheme, the 2016 training set already included 2015 observations, giving the model exposure to the newly prevalent chip-based transaction environment.

These results suggest that the 2015 performance degradation is primarily associated with a **temporal shift in the transaction environment and an elevated false-positive rate**, rather than a failure to recognize fraudulent transactions themselves.

The evidence demonstrates a strong association rather than establishing that the channel shift alone caused the performance degradation.

### 2017 Performance Collapse

Unlike 2015, the 2017 validation period is characterized by an almost complete failure to identify fraudulent transactions.

The model achieved:

- **precision = 0.048**
- **recall = 0.029**
- **F1 = 0.036**

Only **5 of 172 fraudulent transactions** were correctly identified.

The following analysis investigates whether this failure is associated with a change in the characteristics of fraudulent transactions in 2017.